File: 백준/Silver/30506. 가위 가위 가위/가위 가위 가위.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        final int N = 100;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        char[] RSP = new char[N];
        for (int i = 0; i < N; i++) RSP[i] = '2'; // 모두 가위로 초기화

        int K = Integer.parseInt(br.readLine().trim()); // 첫 이긴 판수
        StringBuilder answer = new StringBuilder();

        for (int i = 0; i < N; i++) {
            RSP[i] = '0'; // i번째만 바위
            System.out.println("? " + new String(RSP));
            System.out.flush();
            int result = Integer.parseInt(br.readLine().trim());
            
            if (result == K + 1) {
                answer.append('2'); // 머신은 가위
                K++;
            } else if (result == K) {
                RSP[i] = '5'; // 다음 비교를 위해 보로 바꿔줌
                answer.append('0'); // 머신은 바위
                K++;
            } else if (result == K - 1) {
                RSP[i] = '2'; // 다시 가위로 복구
                answer.append('5'); // 머신은 보
            }
        }
        System.out.println("! " + answer);
        System.out.flush();
    }
}

File: 백준/Silver/30506. 가위 가위 가위/가위 가위 가위.kt
fun main() {
    // 처음에 다 가위냈을 때의 결과 K
    var K = readln().toInt()
    // 처음에 다 가위낸 상태로 시작
    val RSP = "2".repeat(100).toCharArray()
    // 여기에 머신의 패턴을 기록한다
    val answer = StringBuilder()
    repeat(100) {
        // 가위대신 바위내기
        RSP[it] = '0'
        // 결과 확인
        println("? ${RSP.joinToString("")}")
        System.out.flush()
        val result = readln().toInt()

        // 값이 증가 -> 바위로 이길 수 있음 (머신은 가위를 냈음)
        // 값 유지 -> 보로 이길 수 있음 (머신은 바위를 냈음)
        // 값이 감소 -> 가위로 이길 수 있음 (머신은 보를 냈음)
        when(result) {
            K + 1 -> {
                answer.append('2')
                K++
            }
            K -> {
                RSP[it] = '5'
                answer.append('0')
                K++
            }
            K - 1 -> {
                RSP[it] = '2'
                answer.append('5')
            }
        }
    }
    println("! $answer")
    System.out.flush()
}